src/celeritas/ext/ScopedGeantLogger.cc

```diff
@@ -0,0 +1,159 @@
+/*!
+ * Log the actual message.
+ */
+G4int log_impl(G4String const& str, LogLevel level)
+{
+    // Output with dummy file/line
+    ::celeritas::world_logger()({"Geant4", 0}, level) << trim(str);
+
+    // 0 for success, -1 for failure
+    return 0;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Send Geant4 log messages to Celeritas' world logger.
+ */
+class GeantLoggerAdapter : public G4coutDestination
+{
+  public:
+    // Assign to Geant handlers on construction
+    GeantLoggerAdapter();
+    ~GeantLoggerAdapter();
+
+    // Handle error messages
+    G4int ReceiveG4cout(G4String const& str) final;
+    G4int ReceiveG4cerr(G4String const& str) final;
+
+  private:
+#if CELER_G4SSBUF
+    G4coutDestination* saved_cout_{nullptr};
+    G4coutDestination* saved_cerr_{nullptr};
+#endif
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Redirect geant4's stdout/cerr on construction.
+ */
+GeantLoggerAdapter::GeantLoggerAdapter()
+{
+#if CELER_G4SSBUF
+    saved_cout_ = G4coutbuf.GetDestination();
+    saved_cerr_ = G4cerrbuf.GetDestination();
+    G4coutbuf.SetDestination(this);
+    G4cerrbuf.SetDestination(this);
+#else
+    // See Geant4 change global-V11-01-01
+    G4iosSetDestination(this);
+#endif
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Restore iostream buffers on destruction.
+ */
+GeantLoggerAdapter::~GeantLoggerAdapter()
+{
+#if CELER_G4SSBUF
+    G4coutbuf.SetDestination(saved_cout_);
+    G4cerrbuf.SetDestination(saved_cerr_);
+#else
+    G4iosSetDestination(nullptr);
+#endif
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Process stdout message.
+ */
+G4int GeantLoggerAdapter::ReceiveG4cout(G4String const& str)
+{
+    return log_impl(str, LogLevel::diagnostic);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Process stderr message.
+ */
+G4int GeantLoggerAdapter::ReceiveG4cerr(G4String const& str)
+{
+    return log_impl(str, LogLevel::info);
+}
+
+//---------------------------------------------------------------------------//
+//! Global flag for "ownership" of the Geant4 logger
+bool g_adapter_active_{false};
+
+//---------------------------------------------------------------------------//
+}  // namespace
+
+//---------------------------------------------------------------------------//
+/*!
+ * Install the Celeritas Geant4 logger.
+ *
+ * A global flag allows multiple logger adapters to be nested without
+ * consequence.
+ */
+ScopedGeantLogger::ScopedGeantLogger()
+{
+    if (!g_adapter_active_)
+    {
+        static std::mutex capture_mutex;
+        std::lock_guard<std::mutex> scoped_lock{capture_mutex};
+
+        if (!g_adapter_active_)
+        {
+            g_adapter_active_ = true;
+            logger_ = std::make_unique<GeantLoggerAdapter>();
```

Comment:
```suggestion
            logger_ = std::make_unique<GeantLoggerAdapter>();
            g_adapter_active_ = true;
```
With the previous order, you can have another thread that go through this routine with `g_adapter_active_ == true` and `logger_ == nullptr` (or worse, since `logger_` is not an atomic, it "could" be a partially updated value) and thus giving the incorrect behavior (including possibly a crash)

Reply:
Never mind.  `ScopedGeantLogger` is intended to be created on the stack and thus `ScopedGeantLogger::logger_` is not a shared resources.   The code seems correct to implement that "extra" `ScopedGeantLogger` will be ignored.